plugins/kafka/kafka.cpp

```diff
@@ -0,0 +1,1118 @@
+                    }
+                }
+
+                cachedPublishers.clear();
+
+                lockPtr->unlock();
+            }
+
+            /**
+             * Remove previously-created objects that have been inactive
+             * for awhile
+             */
+            void expire()
+            {
+                if (!cachedPublishers.empty())
+                {
+                    // Lock access to avoid thread collisions
+                    lockPtr->lock();
+
+                    time_t      oldestAllowedTime = time(NULL) - OBJECT_EXPIRE_TIMEOUT_SECONDS;
+                    __int32     expireCount = 0;
+
+                    for (ObjMap::iterator x = cachedPublishers.begin(); x != cachedPublishers.end(); /* increment handled explicitly */)
+                    {
+                        // Expire only if the publisher has been inactive and if
+                        // there are no messages in the outbound queue
+                        if (x->second && x->second->getTimeTouched() < oldestAllowedTime && x->second->messagesWaitingInQueue() == 0)
+                        {
+                            // Shutdown the attached poller before deleting
+                            x->second->shutdownPoller();
+
+                            // Delete the object
+                            delete(x->second);
+
+                            // Erase from map
+                            cachedPublishers.erase(x++);
+
+                            ++expireCount;
+                        }
+                        else
+                        {
+                            x++;
+                        }
+                    }
+
+                    lockPtr->unlock();
+
+                    if (traceLevel > 4 && expireCount > 0)
+                    {
+                        DBGLOG("Kafka: Expired %d cached publisher%s", expireCount, (expireCount == 1 ? "" : "s"));
+                    }
+                }
+            }
+
+            /**
+             * Gets an established Publisher, based on unique broker/topic
+             * pairs, or creates a new one.
+             *
+             * @param   _brokers            One or more Kafka brokers, in the
+             *                              format 'name[:port]' where 'name'
+             *                              is either a host name or IP address;
+             *                              multiple brokers can be delimited
+             *                              with commas
+             * @param   _topic              The name of the topic
+             * @param   _pollTimeout        The number of milliseconds to give
+             *                              to librdkafka when executing
+             *                              asynchronous activities
+             *
+             * @return  A pointer to a Publisher* object.
+             */
+            Publisher* getPublisher(const std::string& _brokers, const std::string& _topic, __int32 _pollTimeout)
```

Comment:
style: Many of your parameter names seem to start with an underscore.  We generally only do that inside constructors to differentiate the members from the values that are used to initialise them.